[SM68] Add proposal for Work Graphs (#62)
microsoft/hlsl-specs: proposals/xxxx-work-graphs.md

_Coalescing launch_ nodes represent a thread group operating on a shared array of _input records_. The shader declares the maximum number of records per thread group.

### Records

_Input records_ represent inputs to node shaders, and _output records_ represent outputs from node shaders. Records can be singular or arrays. _Input records_ can be read-only or read-write, while _output records_ are read-write but uninitialized when created.

## Detailed design

### Node Entry Functions

Node shaders are compute shaders built into library targets. Shader entries annotated as `[shader("node")]` are usable as work graph nodes.

A broadcasting launch mode node entry may also be a compute entry (annotated with `[shader("compute")]`). In this case the node's input is an empty input record.

#### Function Attributes

All node shaders except _thread launch_ nodes must specify the thread group size using the `[numthreads(<x>, <y>, <z>)]` attribute.

Node shaders support all compute shader function annotations, but also have new annotations unique to node shaders.

##### **`[NodeLaunch("<mode>")]`**

> Valid values for node launch mode are `broadcasting`, `coalescing`, or `thread`. If the `NodeLaunch` attribute is not specified on a node entry the default launch mode is `broadcasting`.

Author: Note: in historically the attribute name (i.e. `NodeLaunch`) is case-insensitive, but the arguments are case-sensitive. We should follow that with Work Graph attributes.

uninitialized when created.

## Detailed design

### Node Entry Functions

Node shaders are compute shaders built into library targets. Shader entries
annotated as `[shader("node")]` are usable as work graph nodes.

As with all previous shader types, a shader entry function may have only one
shader stage annotation.

#### Function Attributes

All node shaders except _thread launch_ nodes must specify the thread group size
using the `[numthreads(<x>, <y>, <z>)]` attribute.

Node shaders support all compute shader function annotations, but also have new
annotations unique to node shaders.

> Consistent with HLSL grammar attribute names are case-insensitive. Any
> attributes that take string arguments, the string argument values are
> case-sensitive.

##### **`[NodeLaunch("<mode>")]`**
Valid values for node launch mode are `broadcasting`, `coalescing`, or `thread`.
If the `NodeLaunch` attribute is not specified on a node entry the default
launch mode is `broadcasting`.

##### **`[NodeIsProgramEntry]`**

Shaders that can receive input records from both inside and outside the work
graph (i.e. from a command list) require this attribute. Graphs can have more

> than one entry shaders that receive inputs from outside the graph. Nodes which

Author: This attribute is a little weird and the description of it in the D3D spec I didn't find any less confusing. My understanding is that we need to know which nodes are "entry" nodes (meaning they receive records from outside the graph). We make two implicit assumptions:

1) Any node shader that isn't explicitly targeted by an output from another node _is implicitly_ a graph entry node.
2) Any node shader that is explicitly targeted by an output from another node _is not_ a graph entry node.

This attribute is only required for the case where a node is explicitly targeted by an output from another node _and_ the user wants to directly invoke it with records from the command list.

The other sentence in here is that a Work Graph can have more than one entry shader, so there is no limit to the number of implicit or explicit graph entry nodes.

Based on my description here do you have a suggestion for how I can clarify the spec?

read-only and read-write variants:

* `{RW}ThreadNodeInputRecord<RecordTy>` - for thread launch nodes.
* `{RW}DispatchNodeInputRecord<RecordTy>` - for broadcasting launch nodes.
* `{RW}GroupNodeInputRecord<RecordTy>`- for coalescing launch nodes.

The pseudo-HLSL code below describes the basic interface of the
`{RW}{Thread|Dispatch|Group}NodeInputRecord<RecordTy>` classes:

```c++
namespace detail {
///@brief Common interfaces for Thread and Dispatch node input record types.
template <typename RecordTy, bool IsRW> class NodeInputRecordInterface {
  /// @brief Get a copy of the underlying record.
  RecordTy Get() const;

  /// @brief Get a writable reference to the underlying record.
  ///
  /// Only available for non-RW object variants. The non-const Get() returns a
  /// reference to the underlying data.
  std::enable_if_t<IsRW, RecordTy>::type &Get();
};

/// @brief Interface for GroupNodeInputRecord and RWGroupNodeInputRecord
template <typename RecordTy, bool IsRW = false> class GroupNodeInputRecordBase {
  /// @brief Returns the number of records that have been coalesced into the
  /// current thread group.
  ///

>   /// Returns 1..._MaxCount_, where _MaxCount_ is specified by the

Author: There is no guarantee that it a `GroupNodeInputRecord` has `MaxCount` records
when an entry function is called. The guarantee is that there is at least one
record (otherwise the node wouldn't be called), and that there are no more than
defined by the `MaxRecord` attribute.

template <typename RecordTy>
class RWDispatchNodeInputRecord
    : public detail::NodeInputRecordInterface<RecordTy, true> {
  /// @brief Allows thread groups to coordinate reading and writing to a shared
  /// set of records.
  /// @returns Returns `false` for thread groups that are not the last to finish
  /// and `true` for the last thread group to call this method.
  ///
  /// This method must be called by all threads in a dispatch or not called at
  /// all. The call must be in dispatch grid uniform control flow. The callsite
  /// must be uniform across all threads in all thread groups in the dispatch
  /// grid. This method may be called at most once per thread.
  ///
  /// Any violation of these requirements is undefined behavior.
  ///
  /// This method returns `false` for thread groups that are not the last to
  /// finish and these thread groups are not allowed to read or write to the
  /// input. Reading or writing the input after this call on a thread that
  /// returned `false` is undefined behavior.
  ///
  /// This method returns `true` for the last thread group to finish. That
  /// thread group can continue reading and writing to the input.

>   bool FinishedCrossGroupSharing();

Reviewer: That's a lot of text in comments in a pseudo-code block. You'd have more formatting options if you put it outside in metadata. The backticks aren't going to do anything as-is.

Author: The intent for this was that the comments are doxygen formatted (hence the
@<tag> tags). Doxygen does support markdown formatting. These header burbs are
actually complete enough that we could probably generate a doxygen header doc
from them.

<!-- {% raw %} -->

# Work Graphs

* Proposal: [NNNN](NNNN-work-graphs.md)
* Author(s): [Chris Bieneman](https://github.com/llvm-beanz)
* Sponsor: [Greg Roth](https://github.com/pow2clk), [Tex
  Riddell](https://github.com/tex3d)
* Status: **Under Consideration**

> * Planned Version: Shader Model 6.8

Reviewer 2: Shouldn't we update these fields, since really aren't going through approval process for an already-shipped feature.
Also, can we assign a number to this now?  I'd like for the mesh nodes spec to have a number that's higher than this one, since it extends it 😄.

Reviewer: Renumbered and classified.

and DXIL documentation is available in the [Work Graphs Spec on
GitHub](https://microsoft.github.io/DirectX-Specs/d3d/WorkGraphs.html). This
document subsumes that document specifically as it relates to HLSL and the
compiler features.

## Motivation

Current language and runtime limitations make generating work on GPU threads
insufficient to meet the needs of some workloads. If existing GPU features (like
ExecuteIndirect) can't sufficiently generate work the application must generate
the work from the CPU resulting in unnecessary round-tripping between the GPU
and CPU. The Work Graphs feature solves this problem by enabling more robust
GPU-based work creation APIs.

## Proposed solution

The Work Graphs feature allows an application to specify a set of tasks as
_nodes_ in a graph representing a more complex workload. Each _node_ has a fixed
shader which takes one or more _input records_ as input and can produce one or
more _output records_ as output.

### Launch Modes

_Node_ shaders have one of three _launch modes_:

* Thread
* Broadcasting
* Coalescing

_Thread launch_ nodes represent an individual thread of work that processes a

> single _input record_. Thread launch nodes use a thread group size of `(1,1,1)`.

Reviewer: Just FYI, I'm going to reword this a bit. The compiler accepts numthreads on thread nodes if the dimensions are `(1,1,1)`, but it doesn't require it.

  /// @brief Allocate GroupNodeOutputRecords for this thread group.
  /// @returns A handle that collects a set of group node output records.
  /// @param NumRecords The number of records to return.
  ///
  /// GroupNodeOutputRecords are per-group output records. The output record set
  /// is shared across the thread group and the threads work cooperatively to
  /// produce the output records.
  ///
  /// Must be called in thread group uniform control flow. The value of
  /// `NumRecords` and `this` must be uniform across the thread group. If the
  /// value of `NumRecords` is `0`, the object returned is zero-sized and cannot
  /// be indexed.
  ///
  /// This method may not be called from _thread launch_ shaders since they do
  /// not have a thread group.
  GroupNodeOutputRecords<RecordTy> GetGroupNodeOutputRecords(uint NumRecords);

  /// @brief Returns true if the specified output node is in the work graph.
  bool IsValid() const;
};

class EmptyNodeOutput {
  /// @brief Adds `Count` empty output records to the thread node outputs.
  ///
  /// Must be called in thread group uniform control flow.
  void ThreadIncrementOutputCount(uint Count);

>   /// @brief Adds `Count` empty output records to the group node outputs.

Reviewer 2: ```suggestion
  /// @brief Adds `Count` empty output records to the node output, once for the
  /// group, instead of summing the value across threads.
```

`ThreadNodeOutputRecords` and `GroupNodeOutputRecords` objects:

```c++
namespace detail {
template <typename RecordTy> class NodeOutputRecordsBase {
  /// @brief Get a copy of the underlying record.
  RecordTy &Get(uint Index);

  /// @brief Mark the output node as completed.
  ///
  /// Each thread producing an output must call `OutputComplete` at least once.
  /// Calling `OutputComplete()` signals to the runtime that the node output
  /// memory is finalized. The behavior of writes to the output after this call
  /// is undefined.
  ///
  /// Calls to `OutputComplete` must be in thread group uniform control flow
  /// otherwise the behavior is undefined.
  void OutputComplete();
};
} // namespace detail

template <typename RecordTy>
using ThreadNodeOutputRecords = detail::NodeOutputRecordsBase<RecordTy>;

template <typename RecordTy>
using GroupNodeOutputRecords = detail::NodeOutputRecordsBase<RecordTy>;
```

##### Entry Parameter Attributes

> Consistent with HLSL grammar attribute names are case-insensitive. Any
> attributes that take string arguments, the string argument values are
> case-sensitive.

###### **`[MaxRecords(<count>)]`**

> Applies to node inputs in coalescing launch nodes or outputs for any node mode.

Reviewer 2: ```suggestion
Applies to node inputs in coalescing launch or outputs for any launch mode.
```

Required for node inputs for coalescing launch nodes, this attribute restricts
the maximum number of records per thread group. Implementations are not required
to fill to the specified maximum.

When applied to node outputs, this attribute restricts the maximum number of
records produced to the output. When applied to a `NodeOutputArray`, the maximum
applies as the sum of all records across the output array, not per-node.

Node outputs require either the `MaxRecords` or `MaxRecordsSharedWith`
attribute.

> ###### **`[MaxRecordsSharedWith("<name>")]`**
> This attribute applies to node outputs. The named node must have the
> `MaxRecords` attribute applied to its entry. This attribute and the `MaxRecords`
> attribute are mutually exclusive.
> 
> The node that this attribute is applied to shares a maximum record allocation
> with the named node.

Reviewer 2: You specify a parameter here, not a node name.  No quotes.
```suggestion
###### **`[MaxRecordsSharedWith(<parameter>)]`**
This attribute applies to node outputs. The named parameter must have the
`MaxRecords` attribute. This attribute and the `MaxRecords` attribute are
mutually exclusive.

The node output that this attribute is applied to shares a maximum record
allocation with the named node output parameter.
``````

Node outputs require either the `MaxRecords` or `MaxRecordsSharedWith`
attribute.

###### **`[NodeID("<name>", <index> = 0)]`**
This attribute applies to output nodes and defines the name and index of the
output node. If not provided, the default index is 0.

If this attribute is not present on an output, the default node ID for the
output is the name of the parameter, and the index is the default index (0).

###### **`[AllowSparseNodes]`**

> This attribute applies to outputs and allows the work graph to be created even
> if there is not a node defined for the specified output.

Reviewer 2: ```suggestion
This attribute applies to outputs and allows the work graph to be created even
if there is not a node defined for the specified output.  If the output is an
array, each element may or may not have a downstream node defined in the graph.
`IsValid()` can be used to determine whether an output node is defined in the
graph.
```

###### **`[NodeArraySize(<count>)]`**
Specifies the output array size for `NodeOutputArray` or `EmptyNodeOutputArray`
objects.

### New Built-in Functions

#### GetRemainingRecursionLevels

```c++
/// @brief Returns the number of recursion levels remaining against the declared
/// `NodeMaxRecursionDepth`.
///
/// Returns 0 for leaf nodes and if the current node is not recursive.
uint GetRemainingRecursionLevels();
```

For nodes that recurse, the `GetRemainingRecursionLevels()` function returns the
number of remaining recursion levels before reaching the node's maximum
recursion depth.

#### Barrier

```c++

> enum class MEMORY_TYPE_FLAG : uint {

Reviewer 2: In HLSL, `MEMORY_TYPE_FLAG` and `SEMANTIC_FLAG` are not `enum class`, they are like `enum` instead.  They do not require `MEMORY_TYPE_FLAG::` for use.

  UAV_MEMORY = 0x00000001,
  GROUP_SHARED_MEMORY = 0x00000002,
  NODE_INPUT_MEMORY = 0x00000004,
  NODE_OUTPUT_MEMORY = 0x00000008,
  ALL_MEMORY = 0x0000000f,
};

enum class SEMANTIC_FLAG : uint {
  GROUP_SYNC = 0x00000001,
  GROUP_SCOPE = 0x00000002,
  DEVICE_SCOPE = 0x00000004,
};

/// @brief Request a barrier for a set of memory types and/or thread group
/// execution sync.
/// @param MemoryTypeFlags Flag bits as defined by MEMORY_TYPE_FLAG.
/// @param SemanticFlags Flag bits as defined by SEMANTIC_FLAG.
///

> /// `Barrier` must be called from thread group uniform control flow.

Reviewer 2: Barrier only requires group uniform control flow when `GROUP_SYNC` is used.
```suggestion
/// `Barrier` must be called from thread group uniform control flow when
/// `SemanticFlags` includes `GROUP_SYNC`.
```

void Barrier(uint MemoryTypeFlags, uint SemanticFlags);

/// @brief Request a barrier for just the memory used by an object.
/// @param TargetObject The object or resource which owns the memory to apply
/// the barrier to.
/// @param SemanticFlags Flag bits as defined by SEMANTIC_FLAG.
///
/// The TargetObject parameter can be a particular node input/output record
/// object or UAV resource. Groupshared variables are not currently supported.
///

> /// `Barrier` must be called from thread group uniform control flow.

Reviewer 2: Barrier only requires group uniform control flow when `GROUP_SYNC` is used.
```suggestion
/// `Barrier` must be called from thread group uniform control flow when
/// `SemanticFlags` includes `GROUP_SYNC`.
```

void Barrier(Object TargetObject, uint SemanticFlags);
```

Work Graphs introduces a new more flexible implementation of the memory barrier
functions. This function is available in all shader types (including non-node
shaders).

The new `Barrier` function implements a superset of the existing memory barrier
functions which are still supported (i.e. `AllMemoryBarrier{WithGroupSync}()`,
`GroupMemoryBarrier{WithGroupSync}()`, `DeviceMemoryBarrier{WithGroupSync}()`).

In the context of a node shader, `Barrier` enables requesting a memory barrier on
input and/or output record memory specifically, while the implementation is free
to store the data in any memory region.

The pseudo-code below shows implementing the existing HLSL memory barrier
functions using the new `Barrier` function.

```C++

> void AllMemoryBarrier() { Barrier(ALL_MEMORY, DEVICE_SCOPE, 0); }

Reviewer 2: ```suggestion
void AllMemoryBarrier() { Barrier(ALL_MEMORY, DEVICE_SCOPE); }
```

void AllMemoryBarrierWithGroupSync() {
  Barrier(ALL_MEMORY, DEVICE_SCOPE | GROUP_SYNC);
}

void DeviceMemoryBarrier() {

>   Barrier(UAV_MEMORY | GROUP_SHARED_MEMORY, DEVICE_SCOPE, 0);

Reviewer 2: ```suggestion
  Barrier(UAV_MEMORY, DEVICE_SCOPE);
```

}

void DeviceMemoryBarrierWithGroupSync() {

>   Barrier(UAV_MEMORY | GROUP_SHARED_MEMORY, DEVICE_SCOPE | GROUP_SYNC, 0);

Reviewer 2: ```suggestion
  Barrier(UAV_MEMORY, DEVICE_SCOPE | GROUP_SYNC);
```

}

> void GroupMemoryBarrier() { Barrier(GROUP_SHARED_MEMORY, GROUP_SCOPE, 0); }

Reviewer 2: ```suggestion
void GroupMemoryBarrier() { Barrier(GROUP_SHARED_MEMORY, GROUP_SCOPE); }
```

records_ as output.

### Launch Modes

_Node_ shaders have one of three _launch modes_:

* Thread
* Broadcasting
* Coalescing

_Thread launch_ nodes represent an individual thread of work that processes a
single _input record_. Thread launch nodes use a thread group size of `(1,1,1)`.

_Broadcasting launch_ nodes represent a grid of work operating on a single
_input record_. Each input record to a broadcasting node launches a full
dispatch grid. The size of the dispatch grid can either be fixed for the node or
specified in the input.

_Coalescing launch_ nodes represent a thread group operating on a shared array
of _input records_. The shader declares the maximum number of records per thread
group.

### Records

_Input records_ represent inputs to node shaders, and _output records_ represent
outputs from node shaders. Records can be singular or arrays. _Input records_
can be read-only or read-write, while _output records_ are read-write but
uninitialized when created.

## Detailed design

### Node Entry Functions

Node shaders are compute shaders built into library targets. Shader entries
annotated as `[shader("node")]` are usable as work graph nodes.

> A broadcasting launch mode node entry may also be a compute entry (annotated

Author: @pow2clk & @tex3d, I wonder whether or not we should change this. Do we get much benefit from having compute shaders work as node shaders? It wouldn't really be hard to wrap entry points to share shader bodies, and this seems to introduce some unnecessary complication in the language and compiler.

##### **`[NodeDispatchGrid(<x>, <y>, <z>)]`**
Specifies the size of the dispatch grid. Broadcast launch nodes must specify
either the dispatch grid size or maximum dispatch grid size in source. The `x`
`y` and `z` parameters individually cannot exceed (2^16)-1 (65535), and `x*y*z`
cannot exceed (2^24)-1 (16,777,215).

##### **`[NodeMaxDispatchGrid(<x>, <y>, <z>)]`**
Specifies the maximum dispatch grid size when the dispatch grid size is
specified on the input record using the `SV_DispatchGrid` semantic. The `x` `y`
and `z` parameters individually cannot exceed (2^16)-1 (65535), and `x*y*z` cannot
exceed (2^24)-1 (16,777,215).

##### **`[NodeMaxRecursionDepth(<count>)]`**
Specifies the maximum recursion depth for a node. This attribute is required if
one of the outputs of a node is the ID of the node itself.

#### Entry Parameters

Specific types for input parameters depend on the node launch mode, but all node
shaders support two categories of parameters:

* Zero or one [Node Input Parameter](#node-input-objects).
* Zero or more [Node Output Parameter(s)](#node-output-objects).

[Broadcast and Coalescing Launch](launch-modes) shaders also support optional

> system value parameters.

Reviewer 2: I think you misunderstood my previous comment, which I'll resolve since it's pointing to the old text anyway.  I meant that this spec is missing a section describing the allowed system value parameters.

Reviewer: I've lightly modified the corresponding section from the API spec and moved it to its own section after the inputs documentation.

    : public detail::NodeInputRecordInterface<RecordTy, true> {
  /// @brief Allows thread groups to coordinate reading and writing to a shared
  /// set of records.
  /// @returns Returns `false` for thread groups that are not the last to finish
  /// and `true` for the last thread group to call this method.
  ///
  /// This method must be called by all threads in a dispatch or not called at
  /// all. The call must be in dispatch grid uniform control flow. The callsite
  /// must be uniform across all threads in all thread groups in the dispatch
  /// grid. This method may be called at most once per thread.
  ///
  /// Any violation of these requirements is undefined behavior.
  ///
  /// This method returns `false` for thread groups that are not the last to
  /// finish and these thread groups are not allowed to read or write to the
  /// input. Reading or writing the input after this call on a thread that
  /// returned `false` is undefined behavior.
  ///
  /// This method returns `true` for the last thread group to finish. That
  /// thread group can continue reading and writing to the input.
  bool FinishedCrossGroupSharing();
};

template <typename RecordTy>

> using GroupNodeInputRecord = detail::GroupNodeInputRecordsBase<RecordTy, false>;

Reviewer 2: ```suggestion
using GroupNodeInputRecords = detail::GroupNodeInputRecordsBase<RecordTy, false>;
```

template <typename RecordTy>

> using RWGroupNodeInputRecord = detail::GroupNodeInputRecordsBase<RecordTy, true>;

Reviewer 2: ```suggestion
using RWGroupNodeInputRecords = detail::GroupNodeInputRecordsBase<RecordTy, true>;
```
